pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"google.golang.org/grpc"
+)
+
+// PluginInfo contains gRPC connection info
+type PluginInfo struct {
+	SockPath string
+}
+
+// WatchPlugins watches for the creation of plugin sockets at the path
+func WatchPlugins(path string, outChan chan PluginInfo) error {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return err
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	err = watcher.Add(path)
+	if err != nil {
+		watcher.Close()
+		return err
+	}
+
+	go func() {
+		for _, f := range files {
+			outChan <- PluginInfo{path + f.Name()}
+		}
+		for {
+			event := <-watcher.Events
+			if event.Op&fsnotify.Create == fsnotify.Create {
+				outChan <- PluginInfo{event.Name}
+			}
+		}
+	}()
+	return nil
+}
+
+// Dial establishes the gRPC communication with the picked up plugin socket. https://godoc.org/google.golang.org/grpc#Dial
+func Dial(unixSocketPath string) (*grpc.ClientConn, error) {
+	c, err := grpc.Dial(unixSocketPath, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithTimeout(10*time.Second),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}),
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// GetResourceNameFrom returns base resource name from the plugin socket absolute path
+func GetResourceNameFrom(sockPath string) (string, error) {
+	if !((strings.Contains(sockPath, ".sock")) && (strings.Contains(sockPath, "/"))) {
```

Comment:
Do we need the check for "Contains (sockPath, "/")"? May be add a stricter prefix check?

Reply:
This serves for extended resources naming format. Do you have something in mind related to a stricter prefix match?